### PR TITLE
Remove payload field from table sys.segment

### PR DESCRIPTION
**Pull request by yuanlihan**
Screenshot 2020-05-05 at 15 47 27

### Motivation
This PR proposes to remove the `payload` field from table `sys.segments` mainly to handle performance issue caused by serializing object to JSON string.

### Drop alternative
Add a projection rule for scanning table `sys.segment`.

### Operational impact
Since the `payload` field has been already public documented, it's a little aggressive to remove it. But I would like to suppose that there may be rare possibility someone has a dependency on this field. The RESTful `api /druid/coordinator/v1/datasources/{dataSourceName}/segments/{segmentId} ` should be the priority option to retrieve segment metadata.

This PR has:
- [X] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.
<hr>

##### Key changed/added classes in this PR
 * `SystemSchema`

### CODE REVIEW

**Comment by jihoonson** [comment]
Hi @yuanlihan, it seems reasonable to me to not show the whole payload since it has duplicate fields and a garbage (pruned loadSpec). But, I believe other fields such as shardSpec, dimensions, and metrics are still useful. What do you think about adding them instead of payload?

**Comment by yuanlihan** [comment]
> But, I believe other fields such as shardSpec, dimensions, and metrics are still useful. What do you think about adding them instead of payload?

Hi @jihoonson 
I agree that these fields should be useful in some use cases. But I still have some concerns about performance of serializing these fields. Can we add fields when they are needed?
Note that currently almost all sql queries that scan the `sys.segment` table have a sorting clause. And if there are a large number of segments(more than 1 million segments in my case), it will be really slow to load the `Segments` page of the console. And the frontend page already uses the point query API `/druid/coordinator/v1/datasources/{dataSourceName}/segments/{segmentId}` to get the `payload`.

**Comment by jihoonson** [comment]
Oh I meant adding them as new columns but without serialization. For example, we can change [this code](https://github.com/apache/druid/blob/master/sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java#L308-L323) as below:

```java
              return new Object[]{
                  segment.getId(),
                  segment.getDataSource(),
                  segment.getInterval().getStart().toString(),
                  segment.getInterval().getEnd().toString(),
                  segment.getSize(),
                  segment.getVersion(),
                  Long.valueOf(segment.getShardSpec().getPartitionNum()),
                  numReplicas,
                  numRows,
                  IS_PUBLISHED_TRUE, //is_published is true for published segments
                  isAvailable,
                  isRealtime,
                  val.isOvershadowed() ? IS_OVERSHADOWED_TRUE : IS_OVERSHADOWED_FALSE,
                  segment.getShardSpec(),  // new column
                  segment.getDimensions(), // new column
                  segment.getMetrics()     // new column
              };
```

> Can we add these fields when they are needed?

I think we are already using them. The purpose of the "Segments" table or system table is mostly for operators to easily debug their druid cluster. Dimensions and metrics are needed to see if the schema has changed. ShardSpec is needed to see how segments are partitioned so that you can repartition them if needed.

**Comment by yuanlihan** [comment]
Hi @jihoonson, thanks for your suggestion. 

I tried it and it works well for me. And kindly take an another review when you are available. Thanks!

**Comment by surekhasaharan** [comment]
thanks @yuanlihan for raising this PR, i'll also take a look at this soon.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java [hunk 1]
```diff
@@ -333,33 +330,30 @@ public TableType getJdbcTableType()
...
+                numReplicas,
+                val.getValue().getNumRows(),
+                IS_PUBLISHED_FALSE, // is_published is false for unpublished segments
+                // is_available is assumed to be always true for segments announced by historicals or realtime tasks
+                IS_AVAILABLE_TRUE,
+                val.getValue().isRealtime(),
+                IS_OVERSHADOWED_FALSE, // there is an assumption here that unpublished segments are never overshadowed
+                null,
+                null,
+                null
```
Comment (surekhasaharan) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:355]: Why these fields are null, i think available segments would have `shardSpec`, `dimensions` and `metrics` as well

File: web-console/src/views/segments-view/segments-view.tsx [hunk 2]
```diff
@@ -221,7 +221,7 @@ export class SegmentsView extends React.PureComponent<SegmentsViewProps, Segment
           queryParts = compact([
             `SELECT`,
             `  ("start" || '/' || "end") AS "interval",`,
-            `  "segment_id", "datasource", "start", "end", "size", "version", "partition_num", "num_replicas", "num_rows", "is_published", "is_available", "is_realtime", "is_overshadowed", "payload"`,
+            `  "segment_id", "datasource", "start", "end", "size", "version", "partition_num", "num_replicas", "num_rows", "is_published", "is_available", "is_realtime", "is_overshadowed"`,
```
Comment (surekhasaharan) [inline web-console/src/views/segments-view/segments-view.tsx:224]: I don't know TypeScript much, but shouldn't we add the additional fields to this query ?

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java (hunk 1 above)
Comment (surekhasaharan) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:355]: I think you can get the values for those fields by adding them to `AvailableSegmentMetadata` from `DruidSchema` and populating in `PartialSegmentData` in `SystemSchema`

File: web-console/src/views/segments-view/segments-view.tsx (hunk 2 above)
Comment (yuanlihan) [reply web-console/src/views/segments-view/segments-view.tsx:224]: Hi @surekhasaharan, I think the content in this popover already includes these fields. 
image

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java (hunk 1 above)
Comment (yuanlihan) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java]: Hi @surekhasaharan, I have filled these null fields.

File: web-console/src/views/segments-view/segments-view.tsx (hunk 2 above)
Comment (clintropolis) [reply web-console/src/views/segments-view/segments-view.tsx:224]: Yeah, it looks like this information is fetched outside of system table from ```/druid/coordinator/v1/metadata/datasources/${datasourceId}/segments/${segmentId}```

**Pull request merged**